doc: topic blocking vs non-blocking

jrit opened the pull request:
The need for an overview of blocking vs non-blocking was
identified in the docs WG Q1 roadmap. As there are several
topics also pending creation, this one tries to hit the correct
level of detail based on completion of the others.  One
which is referenced is
https://github.com/nodejs/node/pull/4936/files and URLs
within this PR need to change based on where that will land
on the node website.

Any reviews appreciated, thank you.

@chrisdickinson @techjeffharris

jasnell commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+## Blocking
+
+Blocking is when the execution of additional JavaScript in the Node.js process
+must wait until an I/O operation completes. Blocking may occur when using any of
```
nit: it's not just limited to i/o operations

jrit replied:
Would it be more appropriate to just take I/O out of the sentence or is there a better way to explain this? My goal is to distinguish that the process is waiting not *just* because JavaScript is executing.

jasnell replied:
Leaving the I/O in there is fine since that's the *most likely* case. It may just be worthwhile indicating that non-I/O operations can cause it to bog down as well. crypto operations, for instance.

benjamingr replied:
I'm not sure you consider things like waiting for synchronization ( a lock for instance) I/O.

jasnell commented:
generally LGTM. /cc @nodejs/documentation

eljefedelrodeodeljefe commented on doc/topics/blocking-vs-non-blocking.md:
````diff
@@ -0,0 +1,140 @@
+```
+
+The above places a non-blocking call to `unlink` within the callback of
+`readFile` which guarantees the correct order of operations.
````
Could you refer here always to `fs.unlink()`, `fs.readFile()` for consistency?

kelthenoble commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+# Overview of Blocking vs Non-Blocking
+
+This overview covers the difference between blocking and non-blocking calls in
```
I'd prefer to highlight, blocking and non-blocking to `blocking` and `non-blocking` . That way it draws the readers eyes.

eljefedelrodeodeljefe replied:
+1

jasnell replied:
hmm, we usually reserve the code `highlighting` for code or literal values. It may be confusing. Perhaps just use **bold**?

eljefedelrodeodeljefe replied:
Okay. Might be enough.

kelthenoble replied:
@jasnell +1 
<  Perhaps just use bold?

eljefedelrodeodeljefe commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+knowledge of those topics is required. Readers are assumed to have a basic
+understanding of the JavaScript language and Node.js callback pattern.
+
+> In this overview, "I/O" refers primarily to interaction with the system's disk
```
minor nit: avoid duplciation of "This overview". (am no native speaker)

jasnell replied:
+1

eljefedelrodeodeljefe commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+  if (err) throw err;
+  console.log(data);
+});
+moreWork(); // will run before console.log
```
preferably  do `// moreWork()` in order to keep the example runnable. I think this would be more consistent with the rest of the docs.

eljefedelrodeodeljefe commented:
minus nits LGTM

kelthenoble commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+also implement blocking methods, which usually have names that end with `Sync`.
+
+
+## Comparing Code
```
Could we highlight- this is a synchronous file read -> this is a `synchronous` file read 
then highlight: equivalent asynchronous example to equivalent `asynchronous` example
I think it's easier to notice.

kelthenoble commented:
:turtle:  LGTM

Qard commented:
LGTM too, minus the nits.

benjamingr commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+Blocking is when the execution of additional JavaScript in the Node.js process
+must wait until an I/O operation completes. Blocking may occur when using any of
+the synchronous I/O methods in the Node.js standard library that use libuv. Use
+of blocking methods prevents the event loop from doing additional work while
```
_technically_ any native module can do any sort of blocking - it doesn't have to be through libuv but I guess that absolute correctness isn't a goal here.

benjamingr commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+waiting for I/O to complete.
+
+In Node.js, JavaScript that exhibits poor performance due to being CPU intensive
+rather than I/O bound, isn't typically referred to as blocking.
```
There are two notions of what "blocking" is - one is "blocking the event loop" and one is "blocked as in the [waiting thread state](http://www.macdesign.net/capella/it4813/images/Process_states_in_UNIX-color.png)".

When people use "blocking" in node sometimes they mean one and sometimes they mean the other.

benjamingr commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+
+All of the I/O methods in the Node.js standard libraries provide asynchronous
+versions, which are non-blocking, and accept callback functions. Some methods
+also implement blocking methods, which usually have names that end with `Sync`.
```
"Some methods also implement" - the methods don't implement other methods. Maybe "Some methods also have blocking counterparts implemented"?

benjamingr commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+
+In the first example above, `console.log` will be called before `moreWork`. In
+the second example `readFile` is non-blocking so JavaScript execution can
+continue and `moreWork` will be called first. The ability to run `moreWork`
```
question - are we always guaranteed that `readFile` will execute its callback in a following turn of the event loop? I assume yes since it has to hit libuv at least once but I want to be sure.

jrit replied:
I understand that to be the case except for situations where there is an immediate err callback, like when passing `null` as the filename. I think explaining that would be tangential here.

benjamingr commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+
+## Concurrency and Throughput
+
+The Node.js event loop is single threaded, so concurrency refers to the event
```
Maybe it's worth phrasing as JavaScript execution since the event loop itself is not single threaded.

benjamingr commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,140 @@
+## Concurrency and Throughput
+
+The Node.js event loop is single threaded, so concurrency refers to the event
+loop's capacity to resume execution of a JavaScript callback function after
```
Technically it doesn't resume execution - it initiates it. JavaScript always runs synchronously to completion.

jrit commented:
Thank you for the feedback everyone. Aside from specific formatting, the biggest issue seemed to be referring to I/O in a way that wasn't technically accurate. I tried to address that by rewriting some of the `## Blocking` section and referring more directly to "non-JavaScript" instead of "I/O". I hope this will read more accurately without becoming unnecessarily technical.

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+prior knowledge of those topics is required. Readers are assumed to have a
+basic understanding of the JavaScript language and Node.js callback pattern.
+
+> "I/O" refers primarily to interaction with the system's disk and
```
This is quoting, right? Source is missing.

jrit replied:
Not a quote, just a note

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+## Blocking
+
+**Blocking** is when the execution of additional JavaScript in the Node.js
+process must wait until a non-JavaScript operation completes. This happens
```
> non-JavaScript operation

I think this needs to be clarified. For example,

```js
while(1){};
```

will block and it is a JavaScript operation.

jrit replied:
There is a clarification to this a couple sentences later

> In Node.js, JavaScript that exhibits poor performance due to being CPU intensive rather than waiting on a non-JavaScript operation, such as I/O, isn't typically referred to as **blocking**.

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+that use libuv are the most commonly used **blocking** operations. Native
+modules may also have **blocking** methods.
+
+All of the I/O methods in the Node.js standard libraries provide asynchronous
```
standard library, right? Did you mean core modules?

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
````diff
@@ -0,0 +1,143 @@
+});
+```
+
+The first example appears simpler but it has the disadvantage of the second line
````
> simpler

comparative degree used. Shouldn't this be

> The first example appears simpler than the second

or simply

> The first example appears simple

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+the second example `fs.readFile()` is **non-blocking** so JavaScript execution
+can continue and `moreWork()` will be called first. The ability to run
+`moreWork()` without waiting for the file read to complete is a key design
+choice that allows for higher throughout.
```
throughput

jrit replied:
:+1:

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+## Concurrency and Throughput
+
+JavaScript execution in Node.js is single threaded, so concurrency refers to the
+event loop's capacity to execute JavaScript callback functions after completing
```
capacity or capability?

jrit replied:
capacity IMO

YuriSolovyov replied:
Should we mention internal thread pool here?
IMO concurrency in node is more about ability to schedule work to be done and move on, no matter if it is offloaded to internal thread pool or just registered via iocp or linux counterpart, but callbacks are still performed sequentially one-by-one.

jrit replied:
@YuriSolovyov might be better to leave out internal thread pool here unless you think its exclusion makes the statement here false. I think explaining it in another topic where it can receive more depth would be very valuable.

YuriSolovyov replied:
I just wanted to somehow stress that callbacks are still called sequentially and *not* concurrently.
Also, :+1: to more deep and detailed topic about thread pool.

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+event loop's capacity to execute JavaScript callback functions after completing
+other work. Any code that is expected to process requests in a concurrent manner
+depends on the ability of the event loop to continue running as non-JavaScript
+operations like I/O are happening.
```
This sentence has to be rephrased.

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+As an example, let's consider a case where each request to a web server takes
+50ms to complete and 45ms of that 50ms is database I/O that can be done
+asychronously. Choosing **non-blocking** asynchronous operations frees up that
+45ms per request to handle other requests. This is an effective 90% difference
```
Should we specifically mention the percentage gains?

jrit replied:
I feel like the percentage supports the explanation here and makes the impact clearer. Do you feel like it is misleading or have another objection?

thefourtheye replied:
@jrit The percentage would vary from machine to machine and it would not be consistent. I am not sure mentioning the performance gain in percentage here would be a good idea

jasnell replied:
+1 ... if we cannot guarantee that percentage then we should not assert it.

jrit replied:
Some possible resolutions are 
- make a vague statement like "a significant difference", which still implies some percentage change well above zero, but avoids the explicit claim
- completely remove the last sentence of the paragraph to avoid implying any percentage improvement

To the extent this is a hypothetical example in an introductory topic, I favor the first option.

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+The event loop is different than models in many other languages where additional
+threads may be created to handle concurrent work. For an introduction to the
+event loop see [Overview of the Event Loop, Timers, and
+`process.nextTick()`](https://github.com/nodejs/node/pull/4936)
```
Well, PRs are WIP items. So quoting them is a not a good idea I think.

jrit replied:
This shouldn't be merged with this like this, my issue is that there should be a reference to the other topic on nodejs.org but the PR hasn't even been merged into master yet. This was mentioned above when the PR was opened.

thefourtheye commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+fs.readFile('/file.md', (err, data) => {
+  if (err) throw err;
+  console.log(data);
+  fs.unlink('/file.md', err => {
```
Use parens consistently with arrow functions.

tomgco commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+
+All of the I/O methods in the Node.js standard libraries provide asynchronous
+versions, which are **non-blocking**, and accept callback functions. Some
+methods also have **blocking** counterparts, which usually have names that end
```
I believe that they will always be named `Sync` if a method is synchronous.

tomgco commented on doc/topics/blocking-vs-non-blocking.md:
```diff
@@ -0,0 +1,143 @@
+# Overview of Blocking vs Non-Blocking
+
+This overview covers the difference between **blocking** and **non-blocking**
```
Maybe we should clarify around **blocking** being synonymous with synchronous and **non-blocking** with asynchronous as all four terms a used throughout this document.

jrit replied:
Good idea, I'll make the relationship explicit

jrit replied:
@tomgco I put this note at the top of the `Comparing Code` section prior to showing the code samples

techjeffharris commented:
After one read LGTM, sans the link to #4936.  I'll read through it again later to look for nits.  Great Job!

jasnell commented:
Generally LGTM with the nits addressed.

benjamingr commented:
I'm +1 on merging and then improving it later - so LGTM.

laosb commented:
LGTM

jrit commented:
Removed reference to an open PR which was blocking this, it makes more sense to add cross linking at a later date when relevant topics have been merged. Also removed reference to specific percent performance change that had objections from @jasnell and others

This should be in a good place to merge now. Thanks everyone for your feedback.

jasnell commented:
LGTM

Qard commented:
LGTM too.

benjamingr commented:
Landed in https://github.com/nodejs/node/commit/5f1eb434ee750abbd18a25d3b978eb8828f6d1dc - thanks!
